### PR TITLE
Fixes #8045 - Validate domain names per RFC 1123 and allow unicode

### PR DESCRIPTION
**Pull request by tbrisker**
Domain names will be saved to db in punycoded form, and displayed as unicode.

### CODE REVIEW

File: app/validators/domain_name_validator.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class DomainNameValidator < ActiveModel::EachValidator
```
Comment (ohadlevy) [inline app/validators/domain_name_validator.rb:1]: @isratrade @unorthodoxgeek is this the best place to add custom validations? should it be maybe under the model name, e.g. app/validators/domain/name_validator ?

**Comment by dLobatog** [comment]
@tbrisker Why RFC 1035? It doesn't support international domain names unfortunately, in my opinion this should be a very permissive validation that just checks that domains look more or less like a domain (length, dots, etc..), to allow the users to experiment with their own customized domains

**Comment by ohadlevy** [comment]
@eLobato since we are expected to register the domain in DNS, I dont believe we should allow non standard dns names, assuming they will be rejected by DNS ?

**Comment by tbrisker** [comment]
@eLobato @ohadlevy international domain names are actually supposed to be [punycoded](http://en.wikipedia.org/wiki/Punycode) so that they conform to the standards. 
Also domains are case-insensitive which is something we don't check for (ie. duplicates with different capitalization are allowed). 
Should we translate all domain names so that they are lower-cased and punycoded or allow users to create non-standard names?

**Comment by ohadlevy** [comment]
@tbrisker lowercase makes sense to me, as we lowercase the hostname too.
regarding punycoded, I'm not sure, if we end up translating the user input, there might be misconfiguration between actual config files vs data in foreman?

**Comment by tbrisker** [comment]
@ohadlevy I'm not sure I understand what you mean in "misconfiguration between actual config files vs data in foreman"...
Would it make sense to save the puny-coded version in the db (thereby allowing it to pass validation & use the correct value for DNS registration etc.), but display puny-decoded?

File: app/validators/domain_name_validator.rb (hunk 1 above)
Comment (unorthodoxgeek) [inline app/validators/domain_name_validator.rb:1]: I think this is good.

Comment (tbrisker) [inline app/validators/domain_name_validator.rb:1]: I think this validator is also useful in other places, i.e. fqdn validation

**Comment by dLobatog** [comment]
@tbrisker I think that'd be good enough, hopefully it's not complicated to display punycoded addresses with UTF8 characters.

**Comment by tbrisker** [comment]
@eLobato well, it was a bit more complicated then I thought, but looks like it works now (hopefully...)

**Comment by ekohl** [comment]
You should note that this also almost follows [RFC 1123 section 2.1](https://tools.ietf.org/html/rfc1123#page-13) which allows numerical-only hostnames.

File: app/validators/domain_name_validator.rb [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+  # Validate domain name based on RFC 1035, while relaxing the requirement a bit
+  # to only check length and valid characters in case user use a non-standard
+  # domain name.
+  def validate_each(record, attribute, value)
+    record.errors.add(attribute, _("is not a valid domain name")) unless value.downcase =~ /\A[a-z0-9\-]{1,63}(\.[a-z0-9\-]{1,63})*\Z/ && value.length<=255
```
Comment (ekohl) [inline app/validators/domain_name_validator.rb]: For RFC 1123 hostname compliance you should make it `[a-z0-9][a-z0-9\-]{0,62}`.

File: app/models/concerns/scoped_search_extensions.rb [hunk 3]
```diff
@@ -4,7 +4,7 @@ module ScopedSearchExtensions
   module ClassMethods
     def value_to_sql(operator, value)
       return value                 if operator !~ /LIKE/i
-      return value.tr_s('%*', '%') if (value ~ /%|\*/)
+      return value.tr_s('%*', '%') if value =~ /%|\*/
```
Comment (tbrisker) [inline app/models/concerns/scoped_search_extensions.rb]: This was actually silently failing on any "like" search - value is a string and has no method ~

File: app/models/domain.rb [hunk 4]
```diff
@@ -42,6 +46,22 @@ class Domain < ActiveRecord::Base
     end
   }
 
+  def self.complete_for(query, opts = {})
```
Comment (dLobatog) [inline app/models/domain.rb:55]: These two methods are basically duplicated in `Subnet`

File: test/unit/authorizer_test.rb [hunk 5]
```diff
@@ -112,7 +112,7 @@ def setup
     domain1     = FactoryGirl.create(:domain)
     domain2     = FactoryGirl.create(:domain, :name => 'a-domain.to-be-found.com')
     domain3     = FactoryGirl.create(:domain, :name => 'another-domain.to-be-found.com')
-    domain4     = FactoryGirl.create(:domain, :name => 'be_editable.to-be-found.com')
+    domain4     = FactoryGirl.create(:domain, :name => 'be-editable.to-be-found.com')
```
Comment (dLobatog) [inline test/unit/authorizer_test.rb:115]: ["Can subdomains have an underscore in it?"](http://stackoverflow.com/questions/2180465/can-hostname-subdomains-have-an-underscore-in-it)

**Comment by dLobatog** [comment]
[test] because of the latest enabled cops

File: test/unit/authorizer_test.rb (hunk 5 above)
Comment (tbrisker) [inline test/unit/authorizer_test.rb:115]: As far as I can tell (after browsing through quite a few RFCs), underscores in domain names are only used for [SRV](https://tools.ietf.org/html/rfc2782) (A protocol for locating which server supplies a specific service) specifically so they won't conflict with actual hostnames. 
I don't think that in our case we should allow underscores as domain names are used primarily for assigning to hosts & registering to DNS.

**Comment by domcleal** [comment]
Sorry, a Puppet manifest update on our infra caused the database server to be restarted while your tests ran!  [test]

**Comment by tbrisker** [comment]
[test]

**Comment by tbrisker** [comment]
@eLobato How about giving this another look? rebased it into develop.

File: Gemfile [hunk 6]
```diff
@@ -28,6 +28,7 @@ gem 'i18n', '~> 0.6.4'
 gem 'rails-i18n', '~> 3.0.0'
 gem 'turbolinks', '~> 2.5'
 gem 'logging', '>= 1.8.0', '< 3.0.0'
+gem 'simpleidn'
```
Comment (domcleal) [inline Gemfile:31]: This should be pinned, probably on a minor or major version.

**Comment by domcleal** [comment]
1) I'm concerned at all of the search overrides which call to_ascii on the fly, and I think you should consider saving both the ASCII and punycode versions to the DB instead.  This will allow for much faster searching and will significantly reduce the chance of search related bugs - ext_methods don't tend to be unreliable and have a lot of edge cases.

2) The API and UI seem inconsistent, as the API is returning the punycode name, while the UI attempts to show the ASCII name only.  Perhaps both should be exposed as fields - including in the UI, so it's clear that even if you're entering a Unicode name, that it's being converted?

**Pull request closed without merging**